keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1182 @@
+  to migrate their pods.
+
+### Non-Goals
+- Introduce SurgeInterface, which CRDs could implement so that their pods would not be targeted
+  by node drain. Conceptually similar to ScaleInterface (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/scale/interfaces.go).
+- Introduce a field that could include non-critical daemon set pods
+  (priority `system-cluster-critical` or `system-node-critical`) for node maintenance/drain request.
+  The daemon set controller would then gracefully shut down these pods. Critical pods could be
+  overridden by the priority list mentioned below.
+- NodeMaintenance could include a plan of which pods to target first. Similar to graceful node
+  shutdown, we could include a list of priorities to decide which pods should be terminated first.
+  This list could optionally include pod timeouts, but could also wait for all the pods of a given
+  priority class to finish first without a timeout. This could also be used to target daemon set
+  pods of certain priorities (see point above). We could also introduce drain profiles based on
+  these lists. The cluster admin could then choose or create such a profile based on his/her needs.
+  The logic for processing the decision list would be contained in the node maintenance controller,
+  which would set an intent to selected pods to shut down via the TargetedByNodeMaintenance
+  condition.
+- Introduce a node maintenance period or a TTL optional field as an upper bound on the duration of
+  node maintenance. Then the node maintenance would be garbage collected and the node made
+  schedulable again.
+
+## Proposal
+Most of these issues stem from missing a standardized way of detecting a start of the node drain.
+This KEP proposes the introduction of a NodeMaintenance object that would signal an intent to
+gracefully remove pods from a given node. This should complement the node's `.spec.unschedulable`,
+which prevents new pods from being scheduled on such a node. We will focus primarily on
+`kubectl drain`, but the API can also be used by other drain implementations as well (e.g. node
+autoscalers).
+
+We will first introduce the NodeMaintenance API and then later modify the behavior of the
+Kubernetes system to fix all the node drain issues we mentioned earlier. The name, NodeMaintenance
+is subject to change; please see the alternatives section for other suggestions.
+
+To support workload migration, a new controller should be introduced to observe the NodeMaintenance
+objects and then mark pods for migration or removal with a condition. The pods would be selected
+according to the node at first, but the selection mechanism can be extended later. Controllers can
+then implement the migration. The advantage of this approach is that controllers do not have to be
+aware of the NodeMaintenance object (no RBAC changes required). They only have to observe pods they
```

Comment:
Presumably the node would be marked in some fashion as well?

Reply:
I am not sure yet if and how we should do this (condition, annotation). Are you aware of any potential consumers for this?

Reply:
Yes. I had a conversation with a customer right before reading this where the centralized dev ops team did exactly that as a mechanism to notify the application teams, acking that it's not a great mechanism, but it's something that can be automated around. From a cluster admin perspective, it allows you to provide that automation to notify and whatnot.

Reply:
I have added a mention under the node maintenance controller about marking the node.